drivers/digital-io/max14906/iio_max14906.c

```diff
@@ -0,0 +1,639 @@
+			enabled_ch++;
+
+	max14906_iio_channels = no_os_calloc(enabled_ch,
+					     sizeof(*max14906_iio_channels));
+	if (!max14906_iio_channels)
+		return -ENOMEM;
+
+	for (i = 0; i < MAX14906_CHANNELS; i++) {
+		if (!desc->channel_configs[i].enabled ||
+		    desc->channel_configs[i].function == MAX14906_HIGH_Z) {
+			ret = max14906_ch_func(desc->max14906_desc, i, MAX14906_HIGH_Z);
+			if (ret)
+				goto free_channels;
+			continue;
+		}
+
+		max14906_iio_channels[ch_offset] = (struct iio_channel)MAX14906_CHANNEL(i);
+
+		/* Set the direction and attributes based on configuration */
+		if (desc->channel_configs[i].function == MAX14906_IN) {
+			max14906_iio_channels[ch_offset].attributes = max14906_in_attrs;
+			max14906_iio_channels[ch_offset].ch_out = 0;
+		} else {
+			max14906_iio_channels[ch_offset].attributes = max14906_out_attrs;
+			max14906_iio_channels[ch_offset].ch_out = 1;
+		}
+
+		ch_offset++;
+		ret = max14906_ch_func(desc->max14906_desc, i,
+				       desc->channel_configs[i].function);
+		if (ret)
+			goto free_channels;
+	}
+
+	desc->iio_dev->channels = max14906_iio_channels;
+	desc->iio_dev->num_ch = enabled_ch;
+
+	return 0;
+
+free_channels:
+	no_os_free(max14906_iio_channels);
+
+	return ret;
+}
+
+/**
+ * @brief Register read wrapper
+ * @param dev - The iio device structure.
+ * @param reg - The register's address.
+ * @param readval - Register value
+ * @return 0 in case of success, error code otherwise
+ */
+static int max14906_iio_reg_read(struct max14906_iio_desc *dev, uint32_t reg,
+				 uint32_t *readval)
+{
+	return max14906_reg_read(dev->max14906_desc, reg, readval);
+}
+
+/**
+ * @brief Register write wrapper
+ * @param dev - The iio device structure.
+ * @param reg - The register's address.
+ * @param readval - Register value
+ * @return 0 in case of success, error code otherwise
+ */
+static int max14906_iio_reg_write(struct max14906_iio_desc *dev, uint32_t reg,
+				  uint32_t writeval)
+{
+	return max14906_reg_write(dev->max14906_desc, reg, writeval);
+}
+
+/**
+ * @brief Initializes the MAX14906 IIO descriptor.
+ * @param iio_desc - The iio device descriptor.
+ * @param init_param - The structure that contains the device initial parameters.
+ * @return 0 in case of success, an error code otherwise.
+ */
+int max14906_iio_init(struct max14906_iio_desc **iio_desc,
+		      struct max14906_iio_desc_init_param *init_param)
+{
+	struct max14906_iio_desc *descriptor;
+	int ret;
+
+	if (!init_param || !init_param->max14906_init_param)
+		return -EINVAL;
+
+	descriptor = no_os_calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	ret = max14906_init(&descriptor->max14906_desc,
+			    init_param->max14906_init_param);
+	if (ret)
+		goto free_desc;
+
+	descriptor->iio_dev = &max14906_iio_dev;
+
+	memcpy(&descriptor->channel_configs, &init_param->channel_configs,
+	       sizeof(descriptor->channel_configs));
+	ret = max14906_iio_setup_channels(descriptor);
+	if (ret)
+		return ret;
```

Comment:
missing free_desc label, also...max14906_iio_setup_channels is allocation some memory which is should be freed in case of an error

Reply:
the memory allocated by `max14906_iio_setup_channels()` is also freed by the same function, but `max14906_remove()` had to be called indeed.